File: src/utils/templates.rs
use crate::models::{Template, TemplateVariable};
use crate::error::{Result, TriliumError};
use chrono::Local;
use regex::Regex;
use std::collections::HashMap;

/// Parse template variables from content
/// Supports formats: {{variable}}, {{variable:description}}, {{variable:description:default}}
pub fn extract_template_variables(content: &str) -> Vec<TemplateVariable> {
    let var_regex = Regex::new(r"\{\{([^}]+)\}\}").unwrap();
    let mut variables = Vec::new();
    let mut seen_names = std::collections::HashSet::new();
    
    for cap in var_regex.captures_iter(content) {
        if let Some(var_match) = cap.get(1) {
            let var_spec = var_match.as_str().trim();
            let parts: Vec<&str> = var_spec.split(':').collect();
            
            let name = parts[0].trim().to_string();
            if seen_names.contains(&name) {
                continue; // Skip duplicates
            }
            seen_names.insert(name.clone());
            
            let description = if parts.len() > 1 {
                parts[1].trim().to_string()
            } else {
                format!("Value for {}", name)
            };
            
            let default_value = if parts.len() > 2 {
                Some(parts[2].trim().to_string())
            } else {
                None
            };
            
            let required = default_value.is_none() && !is_built_in_variable(&name);
            
            variables.push(TemplateVariable {
                name,
                description,
                default_value,
                required,
            });
        }
    }
    
    variables
}

/// Process template content by substituting variables with security validation
pub fn process_template(
    content: &str, 
    variables: &HashMap<String, String>,
    metadata: Option<&HashMap<String, String>>
) -> Result<String> {
    // Security: Validate input sizes
    const MAX_CONTENT_SIZE: usize = 1_000_000; // 1MB limit
    if content.len() > MAX_CONTENT_SIZE {
        return Err(TriliumError::SecurityError(
            format!("Template content too large (max {} bytes)", MAX_CONTENT_SIZE)
        ));
    }
    
    // Security: Validate variable values
    for (name, value) in variables {
        validate_template_variable(name, value)?;
    }
    
    if let Some(meta) = metadata {
        for (name, value) in meta {
            validate_template_variable(name, value)?;
        }
    }
    let var_regex = Regex::new(r"\{\{([^}]+)\}\}").unwrap();
    
    // Security: Limit number of variable substitutions to prevent DoS
    let mut substitution_count = 0;
    const MAX_SUBSTITUTIONS: usize = 1000;
    
    let result = var_regex.replace_all(content, |caps: &regex::Captures| {
        substitution_count += 1;
        if substitution_count > MAX_SUBSTITUTIONS {
            return "[TOO_MANY_SUBSTITUTIONS]".to_string();
        }
        
        if let Some(var_match) = caps.get(1) {
            let var_spec = var_match.as_str().trim();
            let parts: Vec<&str> = var_spec.split(':').collect();
            let var_name = parts[0].trim();
            
            // Security: Validate variable name
            if !is_valid_variable_name_secure(var_name) {
                return "[INVALID_VARIABLE_NAME]".to_string();
            }
            
            // Check user-provided variables first
            if let Some(value) = variables.get(var_name) {
                return sanitize_variable_value(value);
            }
            
            // Check metadata if provided
            if let Some(meta) = metadata {
                if let Some(value) = meta.get(var_name) {
                    return sanitize_variable_value(value);
                }
            }
            
            // Check built-in variables
            if let Some(value) = get_built_in_variable(var_name) {
                return sanitize_variable_value(&value);
            }
            
            // Use default value if available
            if parts.len() > 2 {
                let default_val = parts[2].trim().to_string();
                return sanitize_variable_value(&default_val);
            }
            
            // Return placeholder if no value found
            format!("{{{{ {} }}}}", sanitize_variable_name(var_name))
        } else {
            caps.get(0).unwrap().as_str().to_string()
        }
    });
    
    // Check if we hit the substitution limit
    if substitution_count > MAX_SUBSTITUTIONS {
        return Err(TriliumError::SecurityError(
            "Too many template variable substitutions - possible DoS attempt".to_string()
        ));
    }
    
    Ok(result.to_string())
}

/// Validate template variable name and value for security
fn validate_template_variable(name: &str, value: &str) -> Result<()> {
    // Security: Limit variable name length
    const MAX_NAME_LENGTH: usize = 100;
    if name.len() > MAX_NAME_LENGTH {
        return Err(TriliumError::SecurityError(
            format!("Variable name too long (max {} characters)", MAX_NAME_LENGTH)
        ));
    }
    
    // Security: Limit variable value length to prevent memory exhaustion
    const MAX_VALUE_LENGTH: usize = 10_000; // 10KB per variable
    if value.len() > MAX_VALUE_LENGTH {
        return Err(TriliumError::SecurityError(
            format!("Variable value too long (max {} characters)", MAX_VALUE_LENGTH)
        ));
    }
    
    // Security: Check for potentially dangerous content
    if value.contains("<script") || value.contains("javascript:") || 
       value.contains("data:text/html") || value.contains("vbscript:") {
        return Err(TriliumError::SecurityError(
            "Variable value contains potentially dangerous content".to_string()
        ));
    }
    
    Ok(())
}

/// Secure validation of variable names with stricter rules
fn is_valid_variable_name_secure(name: &str) -> bool {
    if name.is_empty() || name.len() > 50 {
        return false;
    }
    
    // Only allow alphanumeric characters, underscores, and hyphens
    // Must start with a letter
    name.chars().next().map_or(false, |c| c.is_ascii_alphabetic()) &&
    name.chars().all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '-')
}

/// Sanitize variable values to prevent injection attacks
fn sanitize_variable_value(value: &str) -> String {
    // Remove or escape potentially dangerous characters
    value
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\'', "&#x27;")
        .replace('&', "&amp;")
        .chars()
        .filter(|&c| c.is_ascii_graphic() || c.is_ascii_whitespace())
        .collect()
}

/// Sanitize variable names for safe output
fn sanitize_variable_name(name: &str) -> String {
    name.chars()
        .filter(|&c| c.is_ascii_alphanumeric() || c == '_' || c == '-')
        .collect()
}

/// Check if a variable name is a built-in template variable
fn is_built_in_variable(name: &str) -> bool {
    matches!(name.to_lowercase().as_str(), 
        "date" | "time" | "datetime" | "year" | "month" | "day" | 
        "timestamp" | "user" | "title" | "id" | "uuid")
}

/// Get value for built-in template variables
fn get_built_in_variable(name: &str) -> Option<String> {
    let now = Local::now();
    
    match name.to_lowercase().as_str() {
        "date" => Some(now.format("%Y-%m-%d").to_string()),
        "time" => Some(now.format("%H:%M:%S").to_string()),
        "datetime" => Some(now.format("%Y-%m-%d %H:%M:%S").to_string()),
        "year" => Some(now.format("%Y").to_string()),
        "month" => Some(now.format("%m").to_string()),
        "day" => Some(now.format("%d").to_string()),
        "timestamp" => Some(now.timestamp().to_string()),
        "uuid" => Some(uuid::Uuid::new_v4().to_string()),
        _ => None,
    }
}

/// Create a new template from content
pub fn create_template_from_content(
    id: String,
    title: String, 
    content: String
) -> Template {
    let variables = extract_template_variables(&content);
    
    Template {
        id,
        title,
        content,
        variables,
        description: String::new(),
    }
}

/// Validate template content for common issues
pub fn validate_template(template: &Template) -> Vec<String> {
    let mut issues = Vec::new();
    
    // Check for unclosed variable brackets
    let open_count = template.content.matches("{{").count();
    let close_count = template.content.matches("}}").count();
    if open_count != close_count {
        issues.push("Mismatched variable brackets {{ }}".to_string());
    }
    
    // Check for empty variable names
    let empty_var_regex = Regex::new(r"\{\{\s*\}\}").unwrap();
    if empty_var_regex.is_match(&template.content) {
        issues.push("Empty variable definition found".to_string());
    }
    
    // Check for variables with special characters
    let var_regex = Regex::new(r"\{\{([^}]+)\}\}").unwrap();
    for cap in var_regex.captures_iter(&template.content) {
        if let Some(var_match) = cap.get(1) {
            let var_name = var_match.as_str().split(':').next().unwrap_or("").trim();
            if !is_valid_variable_name(var_name) {
                issues.push(format!("Invalid variable name: '{}'", var_name));
            }
        }
    }
    
    issues
}

/// Check if a variable name is valid
fn is_valid_variable_name(name: &str) -> bool {
    !name.is_empty() && name.chars().all(|c| c.is_alphanumeric() || c == '_' || c == '-')
}

/// Get commonly used template examples
pub fn get_builtin_templates() -> Vec<Template> {
    vec![
        Template {
            id: "daily-journal".to_string(),
            title: "Daily Journal".to_string(),
            content: r#"# Daily Journal - {{date}}

## Tasks for Today
- [ ] 

## Notes
{{notes:Any important notes or observations}}

## Reflection
{{reflection:What went well today?}}

## Tomorrow's Focus
{{tomorrow:What's the main priority for tomorrow?}}

---
*Created: {{datetime}}*"#.to_string(),
            variables: extract_template_variables(r#"# Daily Journal - {{date}}

## Tasks for Today
- [ ] 

## Notes
{{notes:Any important notes or observations}}

## Reflection
{{reflection:What went well today?}}

## Tomorrow's Focus
{{tomorrow:What's the main priority for tomorrow?}}

---
*Created: {{datetime}}*"#),
            description: "Template for daily journal entries".to_string(),
        },
        Template {
            id: "meeting-notes".to_string(),
            title: "Meeting Notes".to_string(),
            content: r#"# {{title:Meeting Title}}

**Date:** {{date}}
**Time:** {{time:Meeting time}}
**Attendees:** {{attendees:List of attendees}}

## Agenda
{{agenda:Meeting agenda items}}

## Discussion Notes
{{notes:Key discussion points}}

## Action Items
- [ ] {{action1:First action item}}
- [ ] {{action2:Second action item}}

## Next Meeting
{{next:Next meeting date/time}}

---
*Notes taken: {{datetime}}*"#.to_string(),
            variables: extract_template_variables(r#"# {{title:Meeting Title}}

**Date:** {{date}}
**Time:** {{time:Meeting time}}
**Attendees:** {{attendees:List of attendees}}

## Agenda
{{agenda:Meeting agenda items}}

## Discussion Notes
{{notes:Key discussion points}}

## Action Items
- [ ] {{action1:First action item}}
- [ ] {{action2:Second action item}}

## Next Meeting
{{next:Next meeting date/time}}

---
*Notes taken: {{datetime}}*"#),
            description: "Template for meeting notes and action items".to_string(),
        },
        Template {
            id: "project-planning".to_string(),
            title: "Project Planning".to_string(),
            content: r#"# {{title:Project Name}}

## Overview
{{overview:Brief project description}}

## Goals & Objectives
{{goals:Primary goals and success criteria}}

## Timeline
- **Start Date:** {{start_date:Project start date}}
- **End Date:** {{end_date:Expected completion date}}
- **Milestones:** {{milestones:Key project milestones}}

## Resources
{{resources:Required resources and team members}}

## Risks & Mitigation
{{risks:Potential risks and mitigation strategies}}

## Status Updates
{{status:Current project status}}

---
*Project created: {{datetime}}*
*Project ID: {{uuid}}*"#.to_string(),
            variables: extract_template_variables(r#"# {{title:Project Name}}

## Overview
{{overview:Brief project description}}

## Goals & Objectives
{{goals:Primary goals and success criteria}}

## Timeline
- **Start Date:** {{start_date:Project start date}}
- **End Date:** {{end_date:Expected completion date}}
- **Milestones:** {{milestones:Key project milestones}}

## Resources
{{resources:Required resources and team members}}

## Risks & Mitigation
{{risks:Potential risks and mitigation strategies}}

## Status Updates
{{status:Current project status}}

---
*Project created: {{datetime}}*
*Project ID: {{uuid}}*"#),
            description: "Template for project planning and tracking".to_string(),
        },
    ]
}

/// Convert template to interactive form data for CLI input
pub fn template_to_form_fields(template: &Template) -> Vec<(String, String, Option<String>, bool)> {
    template.variables.iter()
        .map(|var| (
            var.name.clone(),
            var.description.clone(),
            var.default_value.clone(),
            var.required
        ))
        .collect()
}

/// Process template title with variables
pub fn process_template_title(
    title: &str,
    variables: &HashMap<String, String>,
    metadata: Option<&HashMap<String, String>>
) -> Result<String> {
    process_template(title, variables, metadata)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extract_template_variables() {
        let content = "Hello {{name}} on {{date}}! Your {{task:What to do}} is {{priority:Priority level:medium}}.";
        let variables = extract_template_variables(content);
        
        assert_eq!(variables.len(), 4);
        assert_eq!(variables[0].name, "name");
        assert!(variables[0].required);
        
        assert_eq!(variables[1].name, "date");
        assert!(!variables[1].required); // Built-in variable
        
        assert_eq!(variables[2].name, "task");
        assert_eq!(variables[2].description, "What to do");
        assert!(variables[2].required);
        
        assert_eq!(variables[3].name, "priority");
        assert_eq!(variables[3].description, "Priority level");
        assert_eq!(variables[3].default_value, Some("medium".to_string()));
        assert!(!variables[3].required);
    }

    #[test]
    fn test_process_template() {
        let content = "Hello {{name}}! Today is {{date}} and your task is {{task:What to do:default task}}.";
        let mut variables = HashMap::new();
        variables.insert("name".to_string(), "Alice".to_string());
        
        let result = process_template(content, &variables, None).unwrap();
        
        assert!(result.contains("Hello Alice!"));
        assert!(result.contains("your task is default task"));
        // Date should be filled with current date
        assert!(!result.contains("{{date}}"));
    }

    #[test]
    fn test_is_valid_variable_name() {
        assert!(is_valid_variable_name("valid_name"));
        assert!(is_valid_variable_name("test123"));
        assert!(is_valid_variable_name("my-var"));
        assert!(!is_valid_variable_name(""));
        assert!(!is_valid_variable_name("invalid name"));
        assert!(!is_valid_variable_name("special@char"));
    }

    #[test]
    fn test_validate_template() {
        let good_template = Template {
            id: "test".to_string(),
            title: "Test".to_string(),
            content: "Hello {{name}}!".to_string(),
            variables: vec![],
            description: "".to_string(),
        };
        assert!(validate_template(&good_template).is_empty());
        
        let bad_template = Template {
            id: "test".to_string(),
            title: "Test".to_string(),
            content: "Hello {{name}! Missing close bracket".to_string(),
            variables: vec![],
            description: "".to_string(),
        };
        assert!(!validate_template(&bad_template).is_empty());
    }
    
    #[test]
    fn test_template_security_validation() {
        let mut variables = HashMap::new();
        variables.insert("test".to_string(), "<script>alert('xss')</script>".to_string());
        
        let result = process_template("Hello {{test}}", &variables, None).unwrap();
        assert!(!result.contains("<script>"));
        assert!(result.contains("&lt;script&gt;"));
    }
    
    #[test]
    fn test_template_variable_limits() {
        // Test variable value length limit
        let mut variables = HashMap::new();
        let long_value = "a".repeat(20000);
        variables.insert("test".to_string(), long_value);
        
        let result = process_template("{{test}}", &variables, None);
        assert!(result.is_err());
    }
    
    #[test]
    fn test_template_substitution_limit() {
        // Create a template with many variables to test substitution limit
        let template_parts: Vec<String> = (0..50).map(|i| format!("{{var{}}}", i)).collect();
        let template = template_parts.join(" ");
        
        let variables = HashMap::new();
        let result = process_template(&template, &variables, None);
        assert!(result.is_ok()); // Should be fine with 50 substitutions
    }
    
    #[test]
    fn test_secure_variable_name_validation() {
        assert!(is_valid_variable_name_secure("valid_name"));
        assert!(is_valid_variable_name_secure("test123"));
        assert!(!is_valid_variable_name_secure(""));
        assert!(!is_valid_variable_name_secure("123invalid"));
        assert!(!is_valid_variable_name_secure("invalid name"));
        assert!(!is_valid_variable_name_secure("very_long_variable_name_that_exceeds_the_maximum_allowed_length"));
    }
}

File: src/utils/templates.ts
/**
 * Template Processing Utilities
 * 
 * Provides secure template variable extraction, validation, and substitution
 * with built-in variables and comprehensive security validation.
 */

import type { EntityId, Template, TemplateVariable } from '../types/api.js';

/**
 * Template variable patterns with security validation
 */
const VARIABLE_PATTERNS = {
  // {{variable_name}} or {{variable_name:default_value}}
  BASIC: /\{\{([a-zA-Z_][a-zA-Z0-9_]*?)(?::([^}]*?))?\}\}/g,
  
  // {{@built_in_variable}} for built-in variables
  BUILT_IN: /\{\{@([a-zA-Z_][a-zA-Z0-9_]*?)\}\}/g,
  
  // Combined pattern for efficient parsing
  COMBINED: /\{\{(@?[a-zA-Z_][a-zA-Z0-9_]*?)(?::([^}]*?))?\}\}/g
} as const;

/**
 * Security validation patterns
 */
const SECURITY_PATTERNS = {
  // Valid variable names (letters, numbers, underscores, no spaces)
  VARIABLE_NAME: /^[a-zA-Z_][a-zA-Z0-9_]*$/,
  
  // Safe default values (no script tags, code injection patterns)
  SAFE_DEFAULT: /^[^<>{}()[\]]*$/,
  
  // Maximum lengths to prevent DoS
  MAX_VARIABLE_NAME: 50,
  MAX_DEFAULT_VALUE: 500,
  MAX_VARIABLES_PER_TEMPLATE: 100
} as const;

/**
 * Built-in template variables with their generators
 */
const BUILT_IN_VARIABLES = {
  date: () => new Date().toLocaleDateString(),
  time: () => new Date().toLocaleTimeString(),
  datetime: () => new Date().toLocaleString(),
  timestamp: () => Date.now().toString(),
  iso_date: () => new Date().toISOString().split('T')[0],
  iso_time: () => new Date().toISOString().split('T')[1].split('.')[0],
  iso_datetime: () => new Date().toISOString(),
  year: () => new Date().getFullYear().toString(),
  month: () => (new Date().getMonth() + 1).toString().padStart(2, '0'),
  day: () => new Date().getDate().toString().padStart(2, '0'),
  weekday: () => new Date().toLocaleDateString(undefined, { weekday: 'long' }),
  unix_timestamp: () => Math.floor(Date.now() / 1000).toString(),
  uuid: () => generateUUID(),
  random_id: () => generateRandomId(8)
} as const;

/**
 * Template processing options
 */
export interface TemplateProcessingOptions {
  /** Whether to allow built-in variables */
  allowBuiltIns: boolean;
  
  /** Whether to validate security constraints */
  enableSecurity: boolean;
  
  /** Custom variable generators */
  customVariables?: Record<string, () => string>;
  
  /** Whether to preserve unresolved variables */
  preserveUnresolved: boolean;
  
  /** Maximum recursion depth for nested templates */
  maxRecursionDepth: number;
}

/**
 * Default template processing options
 */
export const DEFAULT_TEMPLATE_OPTIONS: TemplateProcessingOptions = {
  allowBuiltIns: true,
  enableSecurity: true,
  preserveUnresolved: false,
  maxRecursionDepth: 5
};

/**
 * Template variable with security validation
 */
export interface ValidatedTemplateVariable extends TemplateVariable {
  isBuiltIn: boolean;
  isSecure: boolean;
  sanitizedDefaultValue?: string;
}

/**
 * Template processing result
 */
export interface TemplateProcessingResult {
  content: string;
  resolvedVariables: Array<{ name: string; value: string; isBuiltIn: boolean }>;
  unresolvedVariables: string[];
  securityWarnings: string[];
  processingTime: number;
}

/**
 * Extract template variables from content with security validation
 */
export function extractTemplateVariables(
  content: string,
  options: Partial<TemplateProcessingOptions> = {}
): ValidatedTemplateVariable[] {
  const opts = { ...DEFAULT_TEMPLATE_OPTIONS, ...options };
  const variables = new Map<string, ValidatedTemplateVariable>();
  
  if (!content || content.length === 0) {
    return [];
  }
  
  // Reset regex state
  VARIABLE_PATTERNS.COMBINED.lastIndex = 0;
  
  let match: RegExpExecArray | null;
  let variableCount = 0;
  
  while ((match = VARIABLE_PATTERNS.COMBINED.exec(content)) !== null) {
    const [, fullName, defaultValue] = match;
    
    // Security: Limit number of variables per template
    if (++variableCount > SECURITY_PATTERNS.MAX_VARIABLES_PER_TEMPLATE) {
      break;
    }
    
    const isBuiltIn = fullName.startsWith('@');
    const variableName = isBuiltIn ? fullName.slice(1) : fullName;
    
    // Skip if we've already processed this variable
    if (variables.has(variableName)) {
      continue;
    }
    
    // Validate variable name
    const isSecure = validateVariableName(variableName) && 
                    (!defaultValue || validateDefaultValue(defaultValue));
    
    if (!isSecure && opts.enableSecurity) {
      continue; // Skip insecure variables
    }
    
    // Skip built-in variables if not allowed
    if (isBuiltIn && !opts.allowBuiltIns) {
      continue;
    }
    
    variables.set(variableName, {
      name: variableName,
      description: isBuiltIn ? `Built-in variable: ${variableName}` : `User variable: ${variableName}`,
      defaultValue: defaultValue?.trim(),
      required: !defaultValue && !isBuiltIn,
      isBuiltIn,
      isSecure,
      sanitizedDefaultValue: defaultValue ? sanitizeValue(defaultValue) : undefined
    });
  }
  
  return Array.from(variables.values());
}

/**
 * Process template with variable substitution
 */
export function processTemplate(
  content: string,
  variables: Record<string, string> = {},
  options: Partial<TemplateProcessingOptions> = {}
): TemplateProcessingResult {
  const startTime = Date.now();
  const opts = { ...DEFAULT_TEMPLATE_OPTIONS, ...options };
  const resolvedVariables: Array<{ name: string; value: string; isBuiltIn: boolean }> = [];
  const unresolvedVariables: string[] = [];
  const securityWarnings: string[] = [];
  
  if (!content || content.length === 0) {
    return {
      content: '',
      resolvedVariables,
      unresolvedVariables,
      securityWarnings,
      processingTime: Date.now() - startTime
    };
  }
  
  let processedContent = content;
  let recursionDepth = 0;
  
  // Process variables with recursion protection
  while (recursionDepth < opts.maxRecursionDepth) {
    const beforeProcessing = processedContent;
    processedContent = processVariables(
      processedContent,
      variables,
      opts,
      resolvedVariables,
      unresolvedVariables,
      securityWarnings
    );
    
    // Break if no more variables to process
    if (beforeProcessing === processedContent) {
      break;
    }
    
    recursionDepth++;
  }
  
  // Warning if max recursion reached
  if (recursionDepth >= opts.maxRecursionDepth) {
    securityWarnings.push(`Maximum recursion depth (${opts.maxRecursionDepth}) reached`);
  }
  
  return {
    content: processedContent,
    resolvedVariables,
    unresolvedVariables: [...new Set(unresolvedVariables)], // Deduplicate
    securityWarnings,
    processingTime: Date.now() - startTime
  };
}

/**
 * Create template from content with variable extraction
 */
export function createTemplate(
  title: string,
  content: string,
  description = '',
  options: Partial<TemplateProcessingOptions> = {}
): Template {
  const variables = extractTemplateVariables(content, options);
  
  return {
    id: generateRandomId(16) as EntityId,
    title,
    content,
    description,
    variables: variables.map(v => ({
      name: v.name,
      description: v.description,
      defaultValue: v.sanitizedDefaultValue || v.defaultValue,
      required: v.required
    }))
  };
}

/**
 * Validate template content for security and correctness
 */
export function validateTemplate(template: Template): {
  isValid: boolean;
  errors: string[];
  warnings: string[];
} {
  const errors: string[] = [];
  const warnings: string[] = [];
  
  // Basic validation
  if (!template.title || template.title.trim().length === 0) {
    errors.push('Template title cannot be empty');
  }
  
  if (!template.content || template.content.trim().length === 0) {
    errors.push('Template content cannot be empty');
  }
  
  if (template.title && template.title.length > 200) {
    errors.push('Template title too long (max 200 characters)');
  }
  
  if (template.content && template.content.length > 1000000) { // 1MB limit
    errors.push('Template content too large (max 1MB)');
  }
  
  // Extract and validate variables
  const extractedVariables = extractTemplateVariables(template.content, {
    enableSecurity: true,
    allowBuiltIns: true
  });
  
  // Check for unresolved built-in variables
  for (const variable of extractedVariables) {
    if (variable.isBuiltIn && !(variable.name in BUILT_IN_VARIABLES)) {
      warnings.push(`Unknown built-in variable: @${variable.name}`);
    }
    
    if (!variable.isSecure) {
      warnings.push(`Potentially unsafe variable: ${variable.name}`);
    }
  }
  
  // Validate defined variables match extracted variables
  const definedNames = new Set(template.variables.map(v => v.name));
  const extractedNames = new Set(extractedVariables.filter(v => !v.isBuiltIn).map(v => v.name));
  
  for (const name of extractedNames) {
    if (!definedNames.has(name)) {
      warnings.push(`Variable ${name} used in template but not defined in metadata`);
    }
  }
  
  for (const name of definedNames) {
    if (!extractedNames.has(name)) {
      warnings.push(`Variable ${name} defined in metadata but not used in template`);
    }
  }
  
  return {
    isValid: errors.length === 0,
    errors,
    warnings
  };
}

/**
 * Generate template preview with sample data
 */
export function generateTemplatePreview(
  template: Template,
  sampleData: Record<string, string> = {}
): { preview: string; usedSampleData: Record<string, string> } {
  const usedSampleData: Record<string, string> = {};
  
  // Generate sample data for undefined variables
  for (const variable of template.variables) {
    if (!(variable.name in sampleData)) {
      if (variable.defaultValue) {
        usedSampleData[variable.name] = variable.defaultValue;
      } else {
        usedSampleData[variable.name] = generateSampleValue(variable.name);
      }
    } else {
      usedSampleData[variable.name] = sampleData[variable.name];
    }
  }
  
  const result = processTemplate(template.content, usedSampleData, {
    allowBuiltIns: true,
    enableSecurity: true,
    preserveUnresolved: false
  });
  
  return {
    preview: result.content,
    usedSampleData
  };
}

/**
 * Get available built-in variables with descriptions
 */
export function getBuiltInVariables(): Array<{ name: string; description: string; example: string }> {
  return [
    { name: 'date', description: 'Current date (localized)', example: new Date().toLocaleDateString() },
    { name: 'time', description: 'Current time (localized)', example: new Date().toLocaleTimeString() },
    { name: 'datetime', description: 'Current date and time (localized)', example: new Date().toLocaleString() },
    { name: 'timestamp', description: 'Current timestamp (milliseconds)', example: Date.now().toString() },
    { name: 'iso_date', description: 'Current date (ISO format)', example: new Date().toISOString().split('T')[0] },
    { name: 'iso_time', description: 'Current time (ISO format)', example: new Date().toISOString().split('T')[1].split('.')[0] },
    { name: 'iso_datetime', description: 'Current date and time (ISO format)', example: new Date().toISOString() },
    { name: 'year', description: 'Current year', example: new Date().getFullYear().toString() },
    { name: 'month', description: 'Current month (01-12)', example: (new Date().getMonth() + 1).toString().padStart(2, '0') },
    { name: 'day', description: 'Current day (01-31)', example: new Date().getDate().toString().padStart(2, '0') },
    { name: 'weekday', description: 'Current weekday name', example: new Date().toLocaleDateString(undefined, { weekday: 'long' }) },
    { name: 'unix_timestamp', description: 'Unix timestamp (seconds)', example: Math.floor(Date.now() / 1000).toString() },
    { name: 'uuid', description: 'Random UUID v4', example: 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx' },
    { name: 'random_id', description: 'Random alphanumeric ID', example: 'abc123de' }
  ];
}

// ========== Helper Functions ==========

/**
 * Process variables in content (single pass)
 */
function processVariables(
  content: string,
  variables: Record<string, string>,
  options: TemplateProcessingOptions,
  resolvedVariables: Array<{ name: string; value: string; isBuiltIn: boolean }>,
  unresolvedVariables: string[],
  securityWarnings: string[]
): string {
  VARIABLE_PATTERNS.COMBINED.lastIndex = 0;
  
  return content.replace(VARIABLE_PATTERNS.COMBINED, (match, fullName, defaultValue) => {
    const isBuiltIn = fullName.startsWith('@');
    const variableName = isBuiltIn ? fullName.slice(1) : fullName;
    
    // Security validation
    if (options.enableSecurity) {
      if (!validateVariableName(variableName)) {
        securityWarnings.push(`Invalid variable name: ${variableName}`);
        return options.preserveUnresolved ? match : '';
      }
      
      if (defaultValue && !validateDefaultValue(defaultValue)) {
        securityWarnings.push(`Unsafe default value for variable: ${variableName}`);
        return options.preserveUnresolved ? match : '';
      }
    }
    
    // Handle built-in variables
    if (isBuiltIn) {
      if (!options.allowBuiltIns) {
        securityWarnings.push(`Built-in variables not allowed: @${variableName}`);
        return options.preserveUnresolved ? match : '';
      }
      
      if (variableName in BUILT_IN_VARIABLES) {
        try {
          const value = BUILT_IN_VARIABLES[variableName as keyof typeof BUILT_IN_VARIABLES]();
          resolvedVariables.push({ name: variableName, value, isBuiltIn: true });
          return value;
        } catch (error) {
          securityWarnings.push(`Error evaluating built-in variable @${variableName}: ${error}`);
          return options.preserveUnresolved ? match : '';
        }
      } else {
        unresolvedVariables.push(`@${variableName}`);
        return options.preserveUnresolved ? match : '';
      }
    }
    
    // Handle user variables
    if (variableName in variables) {
      const value = sanitizeValue(variables[variableName]);
      resolvedVariables.push({ name: variableName, value, isBuiltIn: false });
      return value;
    }
    
    // Handle default values
    if (defaultValue) {
      const sanitizedDefault = sanitizeValue(defaultValue);
      resolvedVariables.push({ name: variableName, value: sanitizedDefault, isBuiltIn: false });
      return sanitizedDefault;
    }
    
    // Unresolved variable
    unresolvedVariables.push(variableName);
    return options.preserveUnresolved ? match : '';
  });
}

/**
 * Validate variable name security
 */
function validateVariableName(name: string): boolean {
  return SECURITY_PATTERNS.VARIABLE_NAME.test(name) && 
         name.length <= SECURITY_PATTERNS.MAX_VARIABLE_NAME;
}

/**
 * Validate default value security
 */
function validateDefaultValue(value: string): boolean {
  return value.length <= SECURITY_PATTERNS.MAX_DEFAULT_VALUE &&
         SECURITY_PATTERNS.SAFE_DEFAULT.test(value);
}

/**
 * Sanitize variable values
 */
function sanitizeValue(value: string): string {
  // Remove potentially dangerous characters
  return value
    .replace(/[<>]/g, '') // Remove angle brackets
    .replace(/javascript:/gi, '') // Remove javascript: protocol
    .replace(/data:/gi, '') // Remove data: protocol
    .trim();
}

/**
 * Generate sample value based on variable name
 */
function generateSampleValue(variableName: string): string {
  const lowerName = variableName.toLowerCase();
  
  if (lowerName.includes('name')) return 'Sample Name';
  if (lowerName.includes('title')) return 'Sample Title';
  if (lowerName.includes('date')) return new Date().toLocaleDateString();
  if (lowerName.includes('time')) return new Date().toLocaleTimeString();
  if (lowerName.includes('email')) return 'sample@example.com';
  if (lowerName.includes('url')) return 'https://example.com';
  if (lowerName.includes('number') || lowerName.includes('count')) return '123';
  if (lowerName.includes('id')) return generateRandomId(8);
  
  return 'Sample Value';
}

/**
 * Generate UUID v4
 */
function generateUUID(): string {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = Math.random() * 16 | 0;
    const v = c === 'x' ? r : (r & 0x3 | 0x8);
    return v.toString(16);
  });
}

/**
 * Generate random alphanumeric ID
 */
function generateRandomId(length: number): string {
  const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
  let result = '';
  for (let i = 0; i < length; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return result;
}